src/Page/Form.purs

```diff
@@ -177,31 +180,56 @@ component =
         (HE.input $ DropdownQuery field.key)
     Text input ->
       Input.input
-        [ HP.value $ fromMaybe "" $ toString =<< getValue input
+        [ HP.value $ Data.Maybe.fromMaybe "" $ toString =<< getValue input
         , HP.placeholder $ print input.placeholder
         , HP.id_ field.key
+        , HE.onValueChange (HE.input $ UpdateValue field.key <<< UserInput <<< Val <<< String)
         ]
     Toggle input ->
       Toggle.toggle
-        [ HP.checked $ fromMaybe false $ toBoolean =<< getValue input
+        [ HP.checked $ Data.Maybe.fromMaybe false $ toBoolean =<< getValue input
         , HP.id_ field.key
-        , HE.onChecked (HE.input $ UpdateKey field.key <<< Boolean)
+        , HE.onChecked (HE.input $ UpdateValue field.key <<< UserInput <<< Val <<< Boolean)
         ]
 
-eval :: forall m. Query ~> H.ParentDSL State Query (ChildQuery m) ChildSlot Message m
+  renderValidation
+    :: Input ExprType
+    -> Array (H.ParentHTML Query (ChildQuery m) ChildSlot m)
+  renderValidation = case _ of
+    Currency currency -> renderValidation' currency.errors
+    Dropdown dropdown -> renderValidation' dropdown.errors
+    Text text -> renderValidation' text.errors
+    Toggle toggle -> renderValidation' toggle.errors
+    where
+    renderValidation' = errorsToArray >>> case _ of
+      []  -> []
+      [e] -> [ HH.p_ [ HH.text $ renderValidationError e ] ]
+      es  -> [ HH.ul_ $ HH.li_ <<< pure <<< HH.text <<< renderValidationError <$> es ]
+
+    renderValidationError = case _ of
+      Required -> "This field is required"
+      MinLength x -> "Must contain at least " <> show x <> " characters"
+      MaxLength x -> "Cannot contain more than " <> show x <> " characters"
+      InvalidOption x -> x <> " is not a valid option"
+
+eval
+  :: ∀ m
+   . MonadAff m
+  => Query
+  ~> H.ParentDSL State Query (ChildQuery m) ChildSlot Message m
 eval = case _ of
   Initialize a -> a <$ do
     { route } <- H.get
     page' <- case route of
       MVP -> pure (Success mvpPage)
       Profile1 -> pure (Success testPage)
-    bifor_ page' (\e -> H.modify _ { form = Failure e }) \page -> do
-      H.modify_ _ { values = Lynx.Data.Form.keys page }
-      eval (EvalForm page a)
+    bifor_ page'
+      do \e -> H.modify _ { form = Failure e }
+      do \f -> eval (EvalForm f a)
 
   EvalForm expr a -> a <$ do
-    { values } <- H.get
-    let evaled' = Lynx.Data.Form.eval (\key -> Data.Map.lookup key values) expr
+    let values = Lynx.Data.Form.keys expr
```

Comment:
Moved this here, instead of doing it both on `Initialize`, and then updating it from `UpdateValue`. Seemed like an easier way to keep `values` consistent

Reply:
I commented on this here: https://github.com/citizennet/purescript-lynx/pull/16#discussion_r268304532. But I'll pull it here for coherency.

I don't have a preference for how we do things, but `UpdateValue` is now pretty anemic and we don't ever use the `values :: Map Key ExprType` in the `State`. I think if we're going to be recomputing the values every `EvalForm`, let's drop `values` from the `State` and kill `UpdateValue`.